### DISCUSSION TITLE
send feature not working

### DISCUSSION THREAD

**Discussion by ezenkico** [mlua-rs/mlua#372] [Q&A] [answered]
Hi guys

I'm having trouble getting the send safe setup working.

I have the following code.
```rust
async fn run_base(
    input: Vec<u8>
) -> mlua::prelude::LuaResult<()>{
    let lua = Lua::new_with(StdLib::TABLE | StdLib::STRING | StdLib::MATH, LuaOptions::new())?;

    let map_table = lua.create_table()?;
    map_table.set(1, "one")?;
    map_table.set("two", 2)?;

    lua.globals().set("map_table", map_table)?;

    lua.load(input).exec_async().await?;

    Ok(())
}

fn run_lua(
    input: Vec<u8>
) -> Pin<Box<impl Future<Output = mlua::prelude::LuaResult<()>> + Send>>{
    Box::pin(async move{
        run_base(input)
    })
}
```

and this is what I have in cargo

`mlua = {version="0.9.5", features = ["async", "send", "vendored", "luau"]}`

And I'm getting the following errors

`
141 | ) -> Pin<Box<impl Future<Output = mlua::prelude::LuaResult<()>> + Send>>{
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `*mut lua_State` cannot be sent between threads safely
141 | ) -> Pin<Box<impl Future<Output = mlua::prelude::LuaResult<()>> + Send>>{
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `UnsafeCell<mlua::lua::ExtraData>` cannot be shared between threads safely
141  | ) -> Pin<Box<impl Future<Output = mlua::prelude::LuaResult<()>> + Send>>{
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `*const c_void` cannot be sent between threads safely
141  | ) -> Pin<Box<impl Future<Output = mlua::prelude::LuaResult<()>> + Send>>{
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `NonNull<Waker>` cannot be sent between threads safely
     |
141  | ) -> Pin<Box<impl Future<Output = mlua::prelude::LuaResult<()>> + Send>>{
     |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `(dyn for<'a> Fn(&'a Lua) -> Result<LuaVmState, LuaError> + Send + 'static)` cannot be shared between threads safely
141 | ) -> Pin<Box<impl Future<Output = mlua::prelude::LuaResult<()>> + Send>>{
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `*mut c_void` cannot be sent between threads safely
`

The only other possible option I've seen was to use threads and that can get expensive.

If someone can help me out I would appreciate it.

**Comment by khvzak** [comment]
`send` feature works but probably not the way you expect.
You're getting this error becase future returned from `exec_async` is [`!Send`](https://docs.rs/mlua/latest/mlua/struct.Function.html#method.call_async). Making it sendable would require making Lua types `Sync` which is not yet implemented.

Maybe if you tell what you want to achieve I can suggest an alternative solution. Eg. in tokio you can spawn non-Send futures usuing [`LocalSet`](https://docs.rs/tokio/latest/tokio/task/struct.LocalSet.html)

**Reply by ezenkico** [reply]
I appreciate the response.

I can't go into to much detail but in a nutshell I am trying to run lua code on demand and due to other parts of the code I need to do this in an async function that is send safe.  The only thing I can think of is passing things into another thread which will be costly and I would prefer to avoid that.

If you could help that would be great.

Thanks

**Reply by khvzak** [reply]
Even to use `LocalSet` you would need to run another thread. In general passing data to threads is cheap, as all threads share the same address space. If you can spawn a dedicated thread for Lua and then make communication through mpsc channel, it should be sufficient. You can keep references to Lua values using [`RegistryKey`](https://docs.rs/mlua/latest/mlua/struct.RegistryKey.html), they are `Send+Sync`.

**Reply by khvzak** [reply, marked as answer]
An example of using channels
```rust
#[tokio::main]
async fn main() -> mlua::Result<()> {
    let chunk_tx = {
        let (send, mut recv) =
            tokio::sync::mpsc::unbounded_channel::<(String, Sender<LuaResult<()>>)>();
        let rt = Builder::new_current_thread().enable_all().build().unwrap();
        std::thread::spawn(move || {
            rt.block_on(async {
                let lua = Lua::new();
                while let Some((chunk, res_tx)) = recv.recv().await {
                    let res = lua.load(chunk).exec_async().await;
                    res_tx.send(res).unwrap();
                }
            });
        });
        send
    };

    let (res_tx, res_rx) = tokio::sync::oneshot::channel();
    chunk_tx
        .send(("print('Hello, World!')".into(), res_tx))
        .unwrap();
    let res = res_rx.await.unwrap()?;
    println!("Result: {:?}", res);

    Ok(())
}
```

**Reply by ezenkico** [reply]
Yeah, I had a similar idea.  I was just hoping I could avoid it.
If it's not too much trouble, in a future iteration, if you could make it sendable that would be great.  I'll work with this for now though.

I appreciate your help on this.

Thanks.
